cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java

```diff
@@ -0,0 +1,136 @@
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Heartbeat Store that writes heart beat messages and program status messages
+ * to program heartbeat table. This is used for efficiently
+ * scanning and returning results for dashboard status queries
+ */
+public class ProgramHeartbeatDataset extends AbstractDataset {
+  private static final Gson GSON = TriggeringScheduleInfoAdapter.addTypeAdapters(new GsonBuilder()).create();
+  private static final byte[] COLUMN_NAME = Bytes.toBytes("status");
+  private static final DatasetId PROGRAM_HEARTBEAT_INSTANCE_ID =
+    NamespaceId.SYSTEM.dataset(Constants.ProgramHeartbeat.TABLE);
+  private static final byte[] ROW_KEY_SEPARATOR = Bytes.toBytes(":");
+
+  private final Table table;
+
+  public ProgramHeartbeatDataset(Table table) {
+    super("ignored", table);
+    this.table = table;
+  }
+
+  /**
+   * Static method for creating an instance of {@link ProgramHeartbeatDataset}.
+   */
+  public static ProgramHeartbeatDataset getOrCreate(DatasetContext datasetContext, DatasetFramework datasetFramework) {
+    try {
+      Table table = DatasetsUtil.getOrCreateDataset(datasetContext, datasetFramework, PROGRAM_HEARTBEAT_INSTANCE_ID,
+                                                    Table.class.getName(), DatasetProperties.EMPTY);
+      return new ProgramHeartbeatDataset(table);
+    } catch (DatasetManagementException | IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Write {@link RunRecordMeta} to heart beat dataset as value,
+   * with rowKey being <timestamp-in-seconds>:<program-run-id>
+   * @param runRecordMeta row value to write
+   * @param timestampInSeconds used for creating rowKey
+   */
+  public void writeRunRecordMeta(RunRecordMeta runRecordMeta, long timestampInSeconds) {
+    byte[] rowKey = createRowKey(timestampInSeconds, runRecordMeta.getProgramRunId());
+    table.put(rowKey, COLUMN_NAME, Bytes.toBytes(GSON.toJson(runRecordMeta)));
+  }
+
+  private byte[] createRowKey(long timestamp, ProgramRunId programRunId) {
+    return Bytes.add(Bytes.toBytes(timestamp), ROW_KEY_SEPARATOR, Bytes.toBytes(GSON.toJson(programRunId)));
+  }
+
+  /**
+   * Scan the table for the time range and return collection of latest {@link RunRecordMeta}
+   * we maintain the latest {@link RunRecordMeta} identified by {@link ProgramRunId},
+   * Since there can be more than one RunRecordMeta for the
+   * same runId due to multiple state changes and heart beat messages.
+   * @param startRowKey inclusive start rowKey
+   * @param endRowKey exclusive end rowKey
+   * @return collection of {@link RunRecordMeta}
+   */
+  public Collection<RunRecordMeta> scan(byte[] startRowKey, byte[] endRowKey) {
```

Comment:
it's better for this method to take timestamps rather than row keys. Taking rowkeys means the caller is forced to understand how the data is stored.

Reply:
As mentioned in the handler, it would be good to support a method that takes an arbitrary Predicate.